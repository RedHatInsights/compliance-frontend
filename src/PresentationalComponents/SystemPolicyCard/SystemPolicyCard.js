import React from 'react';
import PropTypes from 'prop-types';
import {
  CheckCircleIcon,
  ExclamationCircleIcon,
} from '@patternfly/react-icons';
import {
  Card,
  CardBody,
  CardFooter,
  Text,
  TextContent,
  TextVariants,
  Tooltip,
} from '@patternfly/react-core';
import DateFormat from '@redhat-cloud-services/frontend-components/DateFormat';
import Truncate from '@redhat-cloud-services/frontend-components/Truncate';
import UnsupportedSSGVersion from './UnsupportedSSGVersion';

class SystemPolicyCard extends React.Component {
  complianceIcon = (compliant) =>
    compliant ? (
      <div className="ins-c-policy-card ins-m-compliant">
        <CheckCircleIcon /> Compliant
      </div>
    ) : (
      <div className="ins-c-policy-card ins-m-noncompliant">
        <ExclamationCircleIcon /> Not compliant
      </div>
    );

  fixedPercentage = (value, fixed = 0, withPercent = true) =>
    value.toFixed(fixed) + (withPercent ? '%' : '');

  render() {
    const { policy, style, onClick, isSelected } = this.props;
    const {
      rulesFailed,
      compliant,
      lastScanned,
      score,
      benchmark: { version: ssgVersion },
      supported,
      name,
      policyType,
    } = policy;
    const passedPercentage = this.fixedPercentage(score);
    const truncateDefaults = { expandOnMouseOver: true, hideExpandText: true };

    return (
      <Card
        ouiaId="PolicyCard"
        onClick={(event) => {
          event.stopPropagation();
          onClick(policy);
        }}
        isSelected={isSelected}
        style={style}
      >
        <CardBody>
          <TextContent className="margin-bottom-md">
            <Text
              ouiaId="PolicyCardName"
              className="margin-bottom-top-none"
              component={TextVariants.h4}
            >
              <Truncate text={name} length={110} {...truncateDefaults} />
            </Text>
            <Text
              ouiaId="PolicyCardType"
              style={{ color: 'var(--pf-global--Color--200)' }}
              component={TextVariants.small}
            >
              <Truncate text={policyType} length={110} {...truncateDefaults} />
            </Text>
          </TextContent>
          <div className="margin-bottom-md">
            {supported && this.complianceIcon(compliant)}
            <Text
              ouiaId="PolicyCardFailedRulesScore"
              component={TextVariants.small}
            >
              {rulesFailed} rule{rulesFailed === 1 ? '' : 's'} failed{' '}
              <Tooltip
                position="bottom"
                maxWidth="22em"
                content={
                  'The system compliance score is calculated by OpenSCAP and ' +
                  'is a normalized weighted sum of rules selected for this policy.'
                }
              >
                <span>
                  (Score: {supported ? passedPercentage : 'Unsupported'})
                </span>
              </Tooltip>
            </Text>
          </div>
          <Text className="margin-bottom-none" component={TextVariants.small}>
            <Text ouiaId="PolicyCardSSGVersion">SSG version: {ssgVersion}</Text>
            <Text ouiaId="PolicyCardLastScanned">
              Last scanned:{' '}
              {lastScanned !== 'Never' ? (
                <DateFormat date={Date.parse(lastScanned)} type="relative" />
              ) : (
                lastScanned
              )}
            </Text>
          </Text>
        </CardBody>
        {!supported && (
          <CardFooter style={{ padding: '0' }}>
            <UnsupportedSSGVersion
              ouiaId="PolicyCardUnsupportedSSG"
              ssgVersion={ssgVersion}
              style={{
                paddingTop: 'var(--pf-c-alert--PaddingTop)',
                paddingRight: 'var(--pf-c-card--child--PaddingRight)',
                paddingLeft: 'var(--pf-c-card--child--PaddingLeft)',
                paddingBottom: 'var(--pf-c-alert--PaddingBottom)',
              }}
            />
          </CardFooter>
        )}
      </Card>
    );
  }
}

SystemPolicyCard.propTypes = {
  policy: PropTypes.shape({
    rulesPassed: PropTypes.number,
    rulesFailed: PropTypes.number,
    score: PropTypes.number,
    lastScanned: PropTypes.string,
    refId: PropTypes.string,
    name: PropTypes.string,
    policyType: PropTypes.string,
    compliant: PropTypes.bool,
    benchmark: PropTypes.string,
    supported: PropTypes.bool,
  }),
  style: PropTypes.object,
  onClick: PropTypes.func,
  isSelected: PropTypes.bool,
};

export default SystemPolicyCard;
