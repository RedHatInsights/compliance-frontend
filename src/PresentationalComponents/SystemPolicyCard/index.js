export { default } from './SystemPolicyCard';
export * from './SystemPolicyCard';
