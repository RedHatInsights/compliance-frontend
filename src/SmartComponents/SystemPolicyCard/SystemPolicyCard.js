import React from 'react';
import { CheckCircleIcon, ExclamationCircleIcon } from '@patternfly/react-icons';
import {
    Card,
    CardBody,
    CardFooter,
    Text,
    TextContent,
    TextVariants
} from '@patternfly/react-core';

class SystemPolicyCard extends React.Component {
    constructor(policy) {
        super();
        this.policy = policy.policy;
    }

    complianceIcon(compliant) {
        let result;
        if (compliant) {
            result = <div style={{ fontSize: 'large', color: '#92d400' }} id='policy_compliant'>
                <CheckCircleIcon /> Compliant
            </div>;
        } else {
            result = <div style={{ fontSize: 'large', color: '#a30000' }} id='policy_compliant'>
                <ExclamationCircleIcon/> Noncompliant
            </div>;
        }

        return result;
    }

    render() {
        const totalRules = this.policy.rules_passed + this.policy.rules_failed;
        return (
            <Card>
                <CardBody>
                    <TextContent>
                        <Text style={{ marginBottom: '0px' }} component={TextVariants.small}>External Policy</Text>
                        <Text style={{ marginTop: '0px' }} component={TextVariants.h4}>{this.policy.name}</Text>
                    </TextContent>
                    { this.complianceIcon(this.policy.compliant) }
                    <Text component={TextVariants.small}>
                        { this.policy.rules_passed } of { totalRules } passed
                    </Text>
                    <Text component={TextVariants.medium}>
                        Profile <br/>
                        { this.policy.ref_id }
                    </Text>
                </CardBody>
                <CardFooter>
                    <TextContent>
                        <Text component={TextVariants.small}>
                          Last scanned: { this.policy.last_scanned }
                        </Text>
                    </TextContent>
                </CardFooter>
            </Card>
        );
    };
};

export default SystemPolicyCard;
