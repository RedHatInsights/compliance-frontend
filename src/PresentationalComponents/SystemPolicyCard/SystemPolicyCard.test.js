import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import SystemPolicyCard from './SystemPolicyCard';

describe('SystemPolicyCard component', () => {
  const currentTime = new Date();
  currentTime.setMonth(currentTime.getMonth() - 6);
  const policy = {
    rulesPassed: 30,
    rulesFailed: 10,
    score: 75,
    lastScanned: currentTime.toISOString(),
    refId: 'xccdf_org.ssgproject.content_profile_pci-dss',
    name: 'PCI-DSS Policy',
    policyType: 'PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 7',
    compliant: false,
    supported: true,
    benchmark: { version: '0.1.45' },
    policy: {
      name: 'PCI-DSS Policy',
    },
  };

  it('should render non compliant policy', () => {
    render(<SystemPolicyCard policy={policy} />);

    expect(screen.getByText('Not compliant')).toBeInTheDocument();
  });

  it('should render compliant policy', () => {
    const compliantPolicy = {
      ...policy,
      compliant: true,
    };
    render(<SystemPolicyCard policy={compliantPolicy} />);

    expect(screen.getByText('Compliant')).toBeInTheDocument();
  });

  it('should render an unsupported policy', () => {
    const unsupportedPolicy = {
      ...policy,
      supported: false,
    };
    render(<SystemPolicyCard policy={unsupportedPolicy} />);

    expect(
      screen.getByText('Unsupported SSG version (0.1.45)')
    ).toBeInTheDocument();
  });
});
