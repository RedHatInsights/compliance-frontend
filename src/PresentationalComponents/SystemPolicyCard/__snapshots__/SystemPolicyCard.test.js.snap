// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SystemPolicyCard component should render an unsupported policy 1`] = `
<article
  class="pf-c-card"
  data-ouia-component-id="PolicyCard"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe="true"
  id=""
>
  <div
    class="pf-c-card__body"
  >
    <div
      class="pf-c-content margin-bottom-md"
    >
      <h4
        class="margin-bottom-top-none"
        data-ouia-component-id="PolicyCardName"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        <div
          class="pf-l-stack ins-c-truncate is-block"
        >
          <div
            class="pf-l-stack__item"
          >
            <span
              widget-type="InsightsTruncateBlock"
            >
              PCI-DSS Policy
            </span>
          </div>
        </div>
      </h4>
      <small
        class=""
        data-ouia-component-id="PolicyCardType"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
        style="color:var(--pf-global--Color--200)"
      >
        <div
          class="pf-l-stack ins-c-truncate is-block"
        >
          <div
            class="pf-l-stack__item"
          >
            <span
              widget-type="InsightsTruncateBlock"
            >
              PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 7
            </span>
          </div>
        </div>
      </small>
    </div>
    <div
      class="margin-bottom-md"
    >
      <small
        class=""
        data-ouia-component-id="PolicyCardFailedRulesScore"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        10 rules failed 
        <span>
          (Score: Unsupported)
        </span>
      </small>
    </div>
    <small
      class="margin-bottom-none"
      data-ouia-component-id="OUIA-Generated-Text-10"
      data-ouia-component-type="PF4/Text"
      data-ouia-safe="true"
      data-pf-content="true"
    >
      <p
        class=""
        data-ouia-component-id="PolicyCardSSGVersion"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        SSG version: 0.1.45
      </p>
      <p
        class=""
        data-ouia-component-id="PolicyCardLastScanned"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        Last scanned: 
        <span>
          6 months ago
        </span>
      </p>
    </small>
  </div>
  <div
    class="pf-c-card__footer"
    style="padding:0"
  >
    <div
      aria-label="Warning Alert"
      class="pf-c-alert pf-m-inline pf-m-warning"
      data-ouia-component-id="PolicyCardUnsupportedSSG"
      data-ouia-component-type="PF4/Alert"
      data-ouia-safe="true"
      style="padding-top:var(--pf-c-alert--PaddingTop);padding-right:var(--pf-c-card--child--PaddingRight);padding-left:var(--pf-c-card--child--PaddingLeft);padding-bottom:var(--pf-c-alert--PaddingBottom)"
    >
      <div
        class="pf-c-alert__icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align:-0.125em"
          viewBox="0 0 576 512"
          width="1em"
        >
          <path
            d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          />
        </svg>
      </div>
      <h4
        class="pf-c-alert__title"
      >
        <span
          class="pf-u-screen-reader"
        >
          Warning alert:
        </span>
        Unsupported SSG version (0.1.45)
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="margin-left:.5em;cursor:pointer;color:var(--pf-global--Color--200)"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
          />
        </svg>
      </h4>
    </div>
  </div>
</article>
`;

exports[`SystemPolicyCard component should render policy 1`] = `
<article
  class="pf-c-card"
  data-ouia-component-id="PolicyCard"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe="true"
  id=""
>
  <div
    class="pf-c-card__body"
  >
    <div
      class="pf-c-content margin-bottom-md"
    >
      <h4
        class="margin-bottom-top-none"
        data-ouia-component-id="PolicyCardName"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        <div
          class="pf-l-stack ins-c-truncate is-block"
        >
          <div
            class="pf-l-stack__item"
          >
            <span
              widget-type="InsightsTruncateBlock"
            >
              PCI-DSS Policy
            </span>
          </div>
        </div>
      </h4>
      <small
        class=""
        data-ouia-component-id="PolicyCardType"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
        style="color:var(--pf-global--Color--200)"
      >
        <div
          class="pf-l-stack ins-c-truncate is-block"
        >
          <div
            class="pf-l-stack__item"
          >
            <span
              widget-type="InsightsTruncateBlock"
            >
              PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 7
            </span>
          </div>
        </div>
      </small>
    </div>
    <div
      class="margin-bottom-md"
    >
      <div
        class="ins-c-policy-card ins-m-noncompliant"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align:-0.125em"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          />
        </svg>
         Not compliant
      </div>
      <small
        class=""
        data-ouia-component-id="PolicyCardFailedRulesScore"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        10 rules failed 
        <span>
          (Score: 75%)
        </span>
      </small>
    </div>
    <small
      class="margin-bottom-none"
      data-ouia-component-id="OUIA-Generated-Text-4"
      data-ouia-component-type="PF4/Text"
      data-ouia-safe="true"
      data-pf-content="true"
    >
      <p
        class=""
        data-ouia-component-id="PolicyCardSSGVersion"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        SSG version: 0.1.45
      </p>
      <p
        class=""
        data-ouia-component-id="PolicyCardLastScanned"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        Last scanned: 
        <span>
          6 months ago
        </span>
      </p>
    </small>
  </div>
</article>
`;
